src/VisualStudio/Core/Test.Next/Services/LspDiagnosticsTests.cs

```diff
@@ -0,0 +1,470 @@
+            // 1.  doc1 with id1 = mapped file m1 and id2 = mapped file m2.
+            //
+            // We expect two publish diagnostic notifications ->
+            // 1.  from m1 with id1 (from 1 above).
+            // 2.  from m2 with id2 (from 1 above).
+            var results = await RunPublishDiagnosticsAsync(workspace, diagnosticsMock.Object, expectedNumberOfCallbacks: 2, document).ConfigureAwait(false);
+
+            Assert.Equal(2, results.Count);
+            Assert.Equal("id1", results.Single(p => p.Uri == new Uri(document.FilePath + "m1")).Diagnostics.Single().Code);
+            Assert.Equal("id2", results.Single(p => p.Uri == new Uri(document.FilePath + "m2")).Diagnostics.Single().Code);
+        }
+
+        [Fact]
+        public async Task AddDiagnosticWithMappedFileToManyDocumentsTestAsync()
+        {
+            using var workspace = CreateTestWorkspace(new string[] { "", "" }, out _);
+            var documents = workspace.CurrentSolution.Projects.First().Documents.ToImmutableArray();
+
+            var diagnosticsMock = new Mock<IDiagnosticService>();
+            // Create diagnostic for the first document that has a mapped location.
+            var mappedFilePath = documents[0].FilePath + "m1";
+            var documentOneDiagnostic = await CreateMockDiagnosticDatasWithMappedLocationAsync(documents[0], ("doc1Diagnostic", mappedFilePath)).ConfigureAwait(false);
+            // Create diagnostic for the second document that maps to the same location as the first document diagnostic.
+            var documentTwoDiagnostic = await CreateMockDiagnosticDatasWithMappedLocationAsync(documents[1], ("doc2Diagnostic", mappedFilePath)).ConfigureAwait(false);
+
+            SetupMockWithDiagnostics(diagnosticsMock, documents[0].Id, documentOneDiagnostic);
+            SetupMockWithDiagnostics(diagnosticsMock, documents[1].Id, documentTwoDiagnostic);
+
+            // Publish two document change diagnostic notifications ->
+            // 1.  doc1 with doc1Diagnostic = mapped file m1.
+            // 2.  doc2 with doc2Diagnostic = mapped file m1.
+            //
+            // We expect two publish diagnostic notifications ->
+            // 1.  from m1 with doc1Diagnostic (from 1 above).
+            // 2.  from m1 with doc1Diagnostic and doc2Diagnostic (from 2 above adding doc2Diagnostic to m1).
+            var results = await RunPublishDiagnosticsAsync(workspace, diagnosticsMock.Object, 2, documents[0], documents[1]).ConfigureAwait(false);
+
+            Assert.Equal(2, results.Count);
+            var expectedUri = new Uri(mappedFilePath);
+            Assert.Equal(expectedUri, results[0].Uri);
+            Assert.Equal("doc1Diagnostic", results[0].Diagnostics.Single().Code);
+
+            Assert.Equal(expectedUri, results[1].Uri);
+            Assert.Equal(2, results[1].Diagnostics.Length);
+            Assert.True(results[1].Diagnostics.Contains(d => d.Code == "doc1Diagnostic"));
```

Comment:
There is an Assert.Contains() that takes a lambda you can pass; that way if this fails you don't just get "False".